README.md
Node Stress Suite (NSS) [![Build Status](https://secure.travis-ci.org/Samuel29/NodeStressSuite.png)](http://travis-ci.org/Samuel29/NodeStressSuite)
===

IMPORTANT NOTE
================
Please note that code examples might no longer work. I Will update them later.

this is a brand new fork of NodeLoad.
This project had been idle for a while, so when I needed support for it, I decided to fork ^^

In few words, this fork's main goals are
- to make it compatible with latest versions of NodeJS, to fix HTTP issues when under heavy load.
- add some new reporting features
- eventually add more tools for distributed tests (based on Amazon EC2 instances)
- ...

At present work is still in progress, some examples might not work, and the doc needs to be updated.
You can find at least one working example here : [testLocal.js](https://github.com/Samuel29/NodeStressSuite/tree/master/examples/testLocal.js)


INSTALLING
================

Make sure [node.js](http://nodejs.org/#download) is installed. Then install `nodeload`:

<del>1. Using [npm](http://npmjs.org/) : NOT YET PUBLISHED<del>

1. From source:

        git clone git://github.com/Samuel29/NodeStressSuite.git
        cd NodeStressSuite
        npm link    # optional. enables require('nodeload/<module>') instead of require('./lib/<module>').

2. Or as a single file (this does not install the `nl.js` tool):

        wget https://github.com/Samuel29/NodeStressSuite/raw/master/nss.js

NODELOAD
================

`nodeload` is a collection of independent [node.js](http://nodejs.org/) modules for load testing HTTP services.

As a developer, you should be able to write load tests and get informative reports without having to learn another framework. You should be able to build by example and selectively use the parts of a tool that fit your task. Being a library means that you can use as much or as little of `nodeload` as makes sense, and you can create load tests with the power of a full programming language. For example, if you need to execute some function at a given rate, just use the [`'nodeload/loop'`](https://github.com/Samuel29/nodeload/tree/master/doc/loop.md) module, and write the rest yourself.

In addition, `nodeload` is built for operability. It can always be deployed by simply copying the single file, `nodeload.js`.

Look for examples in the [`examples/`](https://github.com/Samuel29/NodeStressSuite/tree/master/examples) directory and in test cases prefixed with "example" in [`test/`](https://github.com/Samuel29/NodeStressSuite/tree/master/test). Here are simple examples of each module:

### [nl](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/nl.md)

`nl` is an [Apache Bench (ab)](http://httpd.apache.org/docs/2.0/programs/ab.html) like command line tool for running tests quickly. See the [nl documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/nl.md) for details.

    $ examples/test-server.js &    # starts a simple server on port 9000 to load test
    $ ./nl.js -c 10 -n 10000 -i 2 localhost:9000

will send 10,000 queries to http://localhost:9000 using 10 connections. Statistics are printed to the console and graphs can be seen at <http://localhost:8000/>.

### [nodeload](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/nodeload.md)

The `nodeload` module is the primary interface for creating load tests. It includes all the other modules described below, so if you `require('nodeload')`, you don't need to `require()` any of the other ones. Look at the examples in [`examples/loadtesting.ex.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/examples/loadtesting.ex.js) and [`examples/riaktest.ex.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/examples/riaktest.ex.js) or read the [nodeload module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/nodeload.md).

    var nl = require('nodeload');
    var loadtest = nl.run({
        host: 'localhost',
        port: 9000,
        timeLimit: 60,
        targetRps: 500,
        requestGenerator: function(client) {
            var request = client.request('GET', "/" + Math.floor(Math.random()*10000));
            return request;
        }
    });
    loadtest.on('end', function() { console.log('Load test done.'); });

### [remote](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/remote.md)

The `remote` module provides a mechanism for running a distributed load test. See [`examples/remotetesting.ex.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/examples/remotetesting.ex.js) and [`examples/remote.ex.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/examples/remote.ex.js) for examples or read the [remote module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/remote.md).

Start slave instances:

    $ HTTP_PORT=10001 ./nodeload.js  # start a local slave instance on :10001
    $ HTTP_PORT=10002 ./nodeload.js  # start a 2nd slave instance on :10002
    
Create the distributed load test:

    var nl = require('nodeload/remote');
    var cluster = new nl.LoadTestCluster('localhost:8000', ['localhost:8002', 'localhost:8001']);
    cluster.run({
        host: 'localhost',
        port: 9000,
        timeLimit: 60,
        targetRps: 500,
        requestGenerator: function(client) {
            var request = client.request('GET', "/" + Math.floor(Math.random()*10000));
            return request;
        }
    });
    cluster.on('end', function() { console.log('Load test done.'); });

### [stats](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/stats.md)

The `stats` module provides implementations of various statistics objects, like Histograms and Accumulators, and functions, like randomString(), and nextGaussian(). See the [stats module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/stats.md).

    var stats = require('nodeload/stats');
    var histogram = new stats.Histogram();
    for (var i = 0; i < 1000; i++) 
        histogram.put(Math.abs(Math.floor(stats.nextGaussian())));
    console.log('Mean: ' + histogram.mean() + ', 99%: ' + histogram.percentile(0.99));

will output "`Mean: 0.852, 99%: 3`".

### [monitoring](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/monitoring.md)

The `monitoring` module provides a way to track runtime statistics for code that is run concurrently. See [`test/monitoring.test.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/test/monitoring.test.js) for examples or read the [monitoring module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/monitoring.md).

    var monitoring = require('nodeload/monitoring');
    var monitor = new monitoring.Monitor('runtime');
    function asyncFunction() {
        var m = monitor.start();
        setTimeout(function() { m.end(); }, Math.floor(Math.random()*1000));
    }
    for (var i = 0; i < 1000; i++) { asyncFunction(); }
    process.on('exit', function() {
        console.log('Median runtime (ms): ' + monitor.stats['runtime'].percentile(0.5));
    });

will output "`Median runtime (ms): 497`".

### [reporting](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/reporting.md)

The `reporting` module provides a way to graph values over time and present it in a auto-updating HTML page. See [`test/reporting.test.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/test/reporting.test.js) for examples or read the [reporting module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/reporting.md).

    var reporting = require('nodeload/reporting'), 
        stats = require('nodeload/stats'),
        report = reporting.REPORT_MANAGER.addReport('Random Numbers'),
        chart = report.getChart('Gaussian / Pareto vs. Time (minutes)');
    for (var timeout = 0; timeout < 5000; timeout+=500) {
        setTimeout(function() {
                chart.put({
                    'Pareto': stats.nextPareto(0, 100),
                    'Gaussian': stats.nextGaussian()
                });
            }, timeout);
    }

will display a graph on http://localhost:8000/ and save it to an HTML file in the local directory.

### [loop](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/loop.md)

The `loop` module provides a way to execute a function at a set rate and concurrency. See [`test/loop.test.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/test/loop.test.js) for examples or read the [loop module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/loop.md) for details.

    var http = require('http'),
        loop = require('nodeload/loop'),
        requests = 0,
        client = http.createClient(80, 'www.google.com'),
        l = new loop.MultiLoop({
            fun: function(finished) { 
                client.request('GET', '/').end();
                requests++;
                finished();
            },
            rps: 10,
            duration: 3,
            concurrency: 5
        }).start();
    l.on('end', function() { console.log('Total requests: ' + requests) });

will output "`Total requests: 30`".

### [http](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/http.md)

The `http` module provides a generic HTTP server that serves static files and that can be configured with new routes. See [`test/http.test.js`](https://github.com/Samuel29/NodeStressSuite/tree/master/test/http.test.js) for examples or read the [http module documentation](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/http.md).

    var http = require('nodeload/http');
    var server = new http.HttpServer().start(10000);
    server.addRoute('^/hello$', function(url, req, res) {
        res.writeHead(200);
        res.end("Hello");
    });

<http://localhost:8000/package.json> will output the contents of `./package.json`, and <http://localhost:10000/resource> will display "Hello".


CONTRIBUTING
================
File bugs on [github](https://github.com/Samuel29/nodeload/issues), email (any of) the author(s), and fork away. [doc/developers.md](https://github.com/Samuel29/NodeStressSuite/tree/master/doc/developers.md) has brief instructions on getting tests up and running, and will hold more design details in the future. Contributions are always welcome.


RELEASE-NOTES.md
## v0.4.1-BETA (HEAD) ##

* Partially compatible with node v0.8.x :
** this update fixes instability of slaves with sockets hangups and timeouts
** requestLoops now use a custom http.Agent & http.request({agent: myCustomAgent,...}) (instead of deprecated http.createClient / client.request methods)
** TODO: also use this method for master/slave communication (today it raises a "deprecated" warning)
* fix error "Object #<Client> has no method 'destroy'"
* logs/reports files now contain valid characters for windows, and default output dir is now "./results" instead of "."

## v0.4.0 ##

Compatible with node v0.4.x

Features:

* Add graphJmx() and graphProcess(); deprecate spawnAndMonitor(). These provide an easy way to graph JMX attributes as well as output from external processes, such as iostat.
* More readable date strings are used in log files names
* rps is a separate stat from result codes
* Test Results page timestamp does not auto-update on open
* X-axes now use real timestamps rather than minutes since test start

## v0.3.0 (2011/06/16) ##

Compatible with node v0.3.x

Features:

* Add /console/console.html, a jQuery based UI for connecting to multiple nodeload instances simultaneously
* jmxstat/jmxstat.jar allows command line polling of JMX attributes. Combined with reporting.spawnAndMonitor(), Java processes can be monitored during load tests.
* Add 'header-code' statistic which counts number of responses with different values for a given header. For instance, this can be used to graph cache misses/hits from Squid responses using the X-Cache header.

Bug Fixes:

* config: Add 'nodeload/config' module for configuring global parameters
* multiloop: polling time for next change in load or user profiles was always 1 second
* stats: Fix one-off error in Histogram.percentile wouldn't return the greatest number if it is greater than the number of buckets (i.e. in extra[]). Fix Uniques.clear() to actually reset count.
* nl.js: #issue/5: nl.js discarded URL query string and hash

## v0.2.0 (2010/12/01) ##

This release is a substantial, non-backwards-compatible rewrite of nodeload. The major features are:

* [npm](http://npmjs.org/) compatibility
* Independently usable modules: loop, stats, monitoring, http, reporting, and remote
* Addition of load and user profiles

Specific changes to note are:

* npm should be used to build the source

        [~/nodeload]> curl http://npmjs.org/install.sh | sh     # install npm if not already installed
        [~/nodeload]> npm link

* `nodeload` is renamed to `nl` and `nodeloadlib` to `nodeload`.

* addTest() / addRamp() / runTest() is replaced by run():

        var nl = require('nodeload');
        var loadtest = nl.run({ ... test specications ... }, ...);

* remoteTest() / remoteStart() is replaced by LoadTestCluster.run:

        var nl = require('nodeload');
        var cluster = new nl.LoadTestCluster(master:port, [slaves:port, ...]);
        cluster.run({ ... test specifications ...});

* Callbacks and most of the globals (except `HTTP_SERVER` and `REPORT_MANAGER`) have been removed. Instead EventEmitters are used throughout. For example, run() returns an instance of LoadTest, which emits 'update' and 'end' events, replacing the need for both `TEST_MONITOR` and the startTests() callback parameter.

* Scheduler has been replaced by MultiLoop, which also understands load & concurrency profiles.

* Statistics tracking works through event handlers now rather than by wrapping the loop function. See monitoring/monitor.js.

## v0.100.0 (2010/10/06) ##

This release adds nodeloadlib and moves to Dygraph for charting.

## v0.1.0 to v0.1.2 (2010/02/27) ##

Initial releases of nodeload. Tags correspond to node compatible versions. To find a version of node that's compatible with a tag release do `git show <tagname>`.

    For example: git show v0.1.1


lib/loadtesting.js
// ------------------------------------
// Main HTTP load testing interface
// ------------------------------------
//
// This file defines run(), LoadTest, createClient() and extendClient().
//
// This file defines the main API for using nodeload to construct load tests. The main function for 
// starting a load test is run(). Nodeload modules, such as monitoring.js and reporting.js, can also be
// used independently.
//
/*jslint laxbreak: true */
var BUILD_AS_SINGLE_FILE;
if (BUILD_AS_SINGLE_FILE === undefined) {
var http = require('http');
var util = require('./util');
var client = require('./client');
var stats = require('./stats');
var reporting = require('./reporting');
var qputs = util.qputs;
var qprint = util.qprint;
var EventEmitter = require('events').EventEmitter;
var MultiLoop = require('./loop').MultiLoop;
var Monitor = require('./monitoring').Monitor;
var Report = reporting.Report;
var LogFile = stats.LogFile;

var NODELOAD_CONFIG = require('./config').NODELOAD_CONFIG;
var REPORT_MANAGER = reporting.REPORT_MANAGER;
var HTTP_SERVER = require('./http').HTTP_SERVER;
}

/** TEST_OPTIONS defines all of the parameters that can be set in a test specifiction passed to
run(). By default (calling require('nodeload').run({});), will GET localhost:8080/ as fast as possible
with 10 users for 2 minutes. */
var TEST_OPTIONS = {
    name: 'Debug test',                     // A descriptive name for the test

                                            // Specify one of:
    host: 'localhost',                      //   1. (host, port) to connect to via HTTP
    port: 8080,                             //
                                            //
    connectionGenerator: undefined,         //   2. connectionGenerator(), called once for each user. 
                                            //      The return value is passed as-is to requestGenerator,
                                            //      requestLoop, or used internally to generate requests
                                            //      when using (method + path + requestData).
                                            
                                            // Specify one of:
    requestGenerator: undefined,            //   1. requestGenerator: a function
                                            //         function(http.Client) ->  http.ClientRequest
    requestLoop: undefined,                 //   2. requestLoop: is a function
                                            //         function(finished, http.Client)
    method: 'GET',                          //     If must call:
    path: '/',                              //         finished({
    requestData: undefined,                 //             req: http.ClientRequest, 
                                            //             res: http.ClientResponse});
                                            //     after each transaction to finishes to schedule the 
                                            //     next iteration of requestLoop.
                                            //   3. (method + path + requestData) specify a single URL to
                                            //     test
                                            //
    
                                            // Specify one of:
    numUsers: 10,                           //   1. numUsers: number of virtual users concurrently
                                            //      executing therequest loop
    loadProfile: undefined,                 //   2. loadProfile: array with requests/sec over time:
                                            //        [[time (seconds), rps], [time 2, rps], ...]
                                            //      For example, ramp up from 100 to 500 rps and then
                                            //      down to 0 over 20 seconds:
                                            //        [[0, 100], [10, 500], [20, 0]]
                                            
                                            // Specify one of:
    targetRps: Infinity,                    //   1. targetRps: times per second to execute request loop
    userProfile: undefined,                 //   2. userProfile: array with number of users over time:
                                            //        [[time (seconds), # users], [time 2, users], ...]
                                            //      For example, ramp up from 0 to 100 users and back
                                            //      down to 0 over 20 seconds:
                                            //        [[0, 0], [10, 100], [20, 0]]

    numRequests: Infinity,                  // Maximum number of iterations of request loop
    timeLimit: 120,                         // Maximum duration of test in seconds
    delay: 0,                               // Seconds before starting test

    stats: ['latency',                      // Specify list of: 'latency', 'result-codes', 'uniques', 
            'result-codes'],                // 'concurrency', 'http-errors'. These following statistics
                                            // may also be specified with parameters:
                                            //
                                            //     { name: 'latency', percentiles: [0.9, 0.99] }
                                            //     { name: 'http-errors', successCodes: [200,404], log: 'http-errors.log' }
                                            //
                                            // Extend this list of statistics by adding to the
                                            // monitor.js#Monitor.Monitors object.
                                            //
                                            // Note:
                                            // - for 'uniques', traceableRequest() must be used
                                            //   to create the ClientRequest or only 2 will be detected.
};

var LoadTest, generateConnection, requestGeneratorLoop;

/** run(spec, ...) is the primary method for creating and executing load tests with nodeload. See
TEST_OPTIONS for a list of the configuration values in each specification.

@return A LoadTest object with start() / stop() methods, emits 'start' / 'end', and holds statistics
        in .interval and .stats. See LoadTest below.
*/
var run = exports.run = function(specs) {
    specs = (specs instanceof Array) ? specs : util.argarray(arguments);
    var tests = specs.map(function(spec) {
        spec = util.defaults(spec, TEST_OPTIONS);
        var generateRequest = function(client) {
                if (spec.requestGenerator) { return spec.requestGenerator(client); }
                var request = client.request({method:spec.method, path:spec.path, headers:{ 'host': spec.host }});
                if (spec.requestData) {
                    request.write(spec.requestData);
                }
                return request;
            },
            loop = new MultiLoop({
                fun: spec.requestLoop || requestGeneratorLoop(generateRequest),
                argGenerator: spec.connectionGenerator || generateConnection(spec.host, spec.port, !spec.requestLoop),
                concurrencyProfile: spec.userProfile || [[0, spec.numUsers]],
                rpsProfile: spec.loadProfile || [[0, spec.targetRps]],
                duration: spec.timeLimit,
                numberOfTimes: spec.numRequests,
                delay: spec.delay
            }),
            monitor = new Monitor(spec.stats),
            report = new Report(spec.name).updateFromMonitor(monitor);

        loop.on('add', function(loops) { 
            monitor.monitorObjects(loops, 'startiteration', 'enditeration');
        });
        REPORT_MANAGER.addReport(report);
        monitor.name = spec.name;
        monitor.setLoggingEnabled(NODELOAD_CONFIG.LOGS_ENABLED);
        
        return {
            spec: spec,
            loop: loop,
            monitor: monitor,
            report: report,
        };
    });
    
    var loadtest = new LoadTest(tests).start();
    return loadtest;
};

/** LoadTest can be started & stopped. Starting it will fire up the global HTTP_SERVER if it is not
started. Stopping LoadTest will shut HTTP_SERVER down. The expectation is that only one LoadTest instance
is normally running at a time, and when the test finishes, you usually want to let the process end, which
requires stopping HTTP_SERVER. Set loadtest.keepAlive=true to not shut down HTTP_SERVER when done.

LoadTest contains the members:

    - tests: a list of the test objects created by run() from each spec, which contains:
        spec: original specification used by run to create this test object
        loop: a MultiLoop instance that represents all the "vusers" for this job
        monitor: a Monitor instance tracking stats from the MultiLoop instance, loop
        report: a Report which is tracked by REPORT_MANAGER holding a chart for every stat in monitor
    - interval: statistics gathered since the last 'update' event
    - stats: cumulative statistics
    - updateInterval: milliseconds between 'update' events, which includes statistics from the previous 
      interval as well as overall statistics. Defaults to 2 seconds.

LoadTest emits these events:

- 'update', interval, stats: interval has stats since last update. stats contains overall stats.
- 'end': all tests finished

*/
var LoadTest = exports.LoadTest = function LoadTest(tests) {
    EventEmitter.call(this);
    util.PeriodicUpdater.call(this);
    
    var self = this;
    self.tests = tests;
    self.updateInterval = NODELOAD_CONFIG.MONITOR_INTERVAL_MS;
    self.interval = {};
    self.stats = {};
    self.tests.forEach(function(test) {
        self.interval[test.spec.name] = test.monitor.interval;
        self.stats[test.spec.name] = test.monitor.stats;
    });
    self.finishChecker_ = this.checkFinished_.bind(this);
};

util.inherits(LoadTest, EventEmitter);

/** Start running the load test. Starts HTTP_SERVER if it is stopped (unless disabled globally). */
LoadTest.prototype.start = function(keepAlive) {
    var self = this;
    self.keepAlive = keepAlive;

    // clients can catch 'start' event even after calling start().
    process.nextTick(self.emit.bind(self, 'start'));
    self.tests.forEach(function(test) {
        test.loop.start();
        test.loop.on('end', self.finishChecker_);
    });
    
    if (!HTTP_SERVER.running && NODELOAD_CONFIG.HTTP_ENABLED) {
        HTTP_SERVER.start(NODELOAD_CONFIG.HTTP_PORT);
    }

    return self;
};

/** Force the load test to stop. */
LoadTest.prototype.stop = function() {
    this.tests.forEach(function(t) { t.loop.stop(); });
    return this;
};

LoadTest.prototype.update = function() {
    this.emit('update', this.interval, this.stats);
    this.tests.forEach(function(t) { t.monitor.update(); });
    qprint('.');
};

LoadTest.prototype.checkFinished_ = function() {
    if (this.tests.some(function(t) { return t.loop.running; })) { return; }

    this.updateInterval = 0;
    this.update();
    qputs('Done.');

    if (!this.keepAlive) { 
        HTTP_SERVER.stop();
    }

    this.emit('end');
};

/** extendClient extends an existing instance of http.Client by noting the request method and path.
Writing to new requests also emits the 'write' event. This client must be used when using nodeload to
when tracking 'uniques' and 'request-bytes'. */
var extendClient = exports.extendClient = function(client) {
    var wrappedRequest = client.request;
    client.request = function(method, url) {
        var request = wrappedRequest.apply(client, arguments),
            wrappedWrite = request.write,
            wrappedEnd = request.end,
            track = function(data) {
                if (data) {
                    request.emit('write', data);
                    request.body += data.toString();
                }
            };
        request.method = method;
        request.path = url;
        request.body = '';
        request.write = function(data, encoding) {
            track(data);
            return wrappedWrite.apply(request, arguments);
        };
        request.end = function(data, encoding) {
            track(data);
            return wrappedEnd.apply(request, arguments);
        };
        return request;
    };
    return client;
};

/** Same arguments as http.createClient. Returns an extended version of the object (see extendClient) */
//var createClient = exports.createClient = function() {
//    return extendClient(util.createReconnectingClient.apply(this, arguments));
//};

/** Creates a new HTTP connection. This is used as an argGenerator for LoadTest's MultiLoop, so each
"user" gets its own connection. If the load test is using requestGeneratorLoop to generate its requests,
then we also need to terminate pending requests when client errors occur. We emit a fake 'response'
event, so that requestGeneratorLoop can finish its iteration. */
function generateConnection(host, port, detectClientErrors) {
    return function() {
    	var client = new Client(host,port);
        //createClient(port, host);
        if (detectClientErrors) {
            // we need to detect client errors if we're managing the request generation
            client.on('error', function(err) {
                qputs('WARN: Error during HTTP request: ' + (err ? err.toString() : 'unknown'));
            });
//            client.on('reconnect', function(oldclient) {
//                // For each pending outgoing request, simulate an empty response
//                if (oldclient._outgoing) {
//                    oldclient._outgoing.forEach(function(req) {
//                        if (req instanceof http.ClientRequest) {
//                            req.emit('response', new EventEmitter());
//                        }
//                    });
//                }
//            });
        }
        return client;
    };
}

/** Wrapper for request generator function, generator

@param generator A function:

                     function(http.Client) -> http.ClientRequest

                 The http.Client is provided by nodeload. The http.ClientRequest may contain an extra
                 .timeout field specifying the maximum milliseconds to wait for a response.

@return A Loop compatible function, function(loopFun, http.Client). Each iteration makes an HTTP
        request by calling generator. loopFun({req: http.ClientRequest, res: http.ClientResponse}) is
        called when the HTTP response is received or the request times out. */
function requestGeneratorLoop(generator) {
    return function(finished, client) {
        var running = true, timeoutId, request = generator(client);
        var callFinished = function(response) {
            if (running) { 
                running = false;
                clearTimeout(timeoutId);
                response.statusCode = response.statusCode || 0;
                finished({req: request, res: response});
            }
        };
        if (request) {
            if (request.timeout > 0) {
                timeoutId = setTimeout(function() {
                				qputs("request timed out");
                                callFinished(new EventEmitter());
                            }, request.timeout);
            }
            request.on('response', function(response) {
                callFinished(response);
            });
            request.end();
        } else {
            finished(null);
        }
    };
}